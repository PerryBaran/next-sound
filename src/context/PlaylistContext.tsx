'use client'

import { createContext, useContext, useState } from 'react'

interface Song {
  image?: string
  artistName: string
  albumName: string
  songName: string
  audio: string
}

interface Playlist extends Song {
  key: string
}

const PlaylistContext = createContext({
  playlist: [] as Playlist[],
  playlistIndex: 0,
  skipSong: (i: number | boolean) => {},
  handleAddToPlaylist: (
    songs: Song[],
    addNext: boolean,
    playNow: boolean
  ) => {},
  removeFromPlaylist: (i: number) => {},
  playing: false,
  handlePlaying: (isPlaying: boolean | undefined) => {}
})

const usePlaylistContext = () => useContext(PlaylistContext)

function PlaylistProvider({ children }: { children: React.ReactNode }) {
  const [playlist, setPlaylist] = useState<Playlist[]>([])
  const [playlistIndex, setPlaylistIndex] = useState(0)
  const [playing, setPlaying] = useState(false)

  const addToPlaylist = (songs: Song[], addNext = false) => {
    const data = songs.map((song) => {
      return { ...song, key: crypto.randomUUID() }
    })

    setPlaylist((prev) => {
      if (addNext) {
        const clone = [...prev]
        clone.splice(playlistIndex + 1, 0, ...data)
        return clone
      }
      return [...prev, ...data]
    })
  }

  const handlePlayNow = () => {
    if (playlist.length !== 0) {
      setPlaylistIndex((prev) => prev + 1)
    }
    setPlaying(true)
  }

  const handleAddToPlaylist = (
    songs: Song[],
    addNext: boolean,
    playNow: boolean
  ) => {
    addToPlaylist(songs, addNext)
    if (playNow) {
      handlePlayNow()
    }
  }

  const removeFromPlaylist = (i: number) => {
    setPlaylist((prev) => {
      const clone = [...prev]
      clone.splice(i, 1)
      return clone
    })

    setPlaylistIndex((prev) => {
      if (i > prev || prev === 0) {
        return prev
      }
      return prev - 1
    })
  }

  const skipSong = (value: number | boolean) => {
    const lastPlaylistIndex = playlist.length - 1
    if (typeof value === 'number') {
      if (value > lastPlaylistIndex) {
        setPlaylistIndex(lastPlaylistIndex)
      } else if (value < 0) {
        setPlaylistIndex(0)
      } else {
        setPlaylistIndex(value)
      }
    } else if (value) {
      if (playlistIndex === lastPlaylistIndex) {
        setPlaylistIndex(0)
      } else {
        setPlaylistIndex((prev) => prev + 1)
      }
    } else {
      if (playlistIndex > 0) {
        setPlaylistIndex((prev) => prev - 1)
      } else {
        setPlaylistIndex(lastPlaylistIndex)
      }
    }
  }

  const handlePlaying = (isPlaying: boolean | undefined) => {
    if (isPlaying === undefined) {
      setPlaying((prev) => !prev)
    } else {
      setPlaying(isPlaying)
    }
  }

  return (
    <PlaylistContext.Provider
      value={{
        playlist,
        playlistIndex,
        handleAddToPlaylist,
        removeFromPlaylist,
        skipSong,
        playing,
        handlePlaying
      }}
    >
      {children}
    </PlaylistContext.Provider>
  )
}

export { usePlaylistContext, PlaylistProvider }
