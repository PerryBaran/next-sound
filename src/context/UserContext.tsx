'use client'

import { createContext, useContext, useState, useEffect } from 'react'
import { getUserById } from '@/requests/users'
import Cookie from 'js-cookie'
import jwtDecode from 'jwt-decode'

interface User {
  name?: string
  id?: string
}

const emptyUser = {
  name: '',
  id: ''
}

const UserContext = createContext({
  user: emptyUser,
  handleLogin: (data: User | undefined) => {}
})

const useUserContext = () => useContext(UserContext)

function UserProvider({ children }: { children: React.ReactNode }) {
  const [user, setUser] = useState(emptyUser)

  const handleLogin = (data: User | undefined) => {
    if (data === undefined) {
      Cookie.remove('userToken')
      setUser(emptyUser)
    } else {
      setUser((prev) => {
        return { ...prev, ...data }
      })
    }
  }

  useEffect(() => {
    ;(async () => {
      const token = Cookie.get('userToken')
      if (!token) return
      const { id }: { id: string } = jwtDecode(token)
      const response = await getUserById(id)
      handleLogin({
        name: response.name,
        id
      })
    })()
  }, [])

  return (
    <UserContext.Provider value={{ user, handleLogin }}>
      {children}
    </UserContext.Provider>
  )
}

export { useUserContext, UserProvider }
