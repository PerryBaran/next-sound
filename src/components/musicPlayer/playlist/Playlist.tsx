'use client'

import Image from 'next/image'
import { usePlaylistContext } from '@/context/PlaylistContext'
import css from './playlist.module.css'
import { play, music } from '@/media/icons'
import { useState } from 'react'
import { menu } from '@/media/icons'

interface Props {
  handlePlaying: (isPlayin: boolean | undefined) => void
}

export default function Playlist({ handlePlaying }: Props) {
  const { playlist, playlistIndex, skipSong, removeFromPlaylist } =
    usePlaylistContext()
  const [showPlaylist, setShowPlaylist] = useState(false)

  const handleSetPlaying = (i: number) => {
    skipSong(i)
    handlePlaying(true)
  }

  return (
    <>
      {showPlaylist && (
        <div className={css.container}>
          {playlist.map((song, i) => {
            return (
              <div
                key={song.key}
                className={`${css.playlist__song} 
                ${playlistIndex === i && css.play__song_active}
              }`}
              >
                <div className={css.playlist__song_left}>
                  <Image
                    src={song.image || music}
                    alt={`${song.albumName} cover art`}
                    height={30}
                    width={30}
                  />
                </div>
                <div className={css.playlist__song_right}>
                  <div className={css.playlist__song_info}>
                    <p className={css.playlist__song_artist}>
                      {song.artistName}
                    </p>
                    <p className={css.playlist__song_name}>{song.songName}</p>
                  </div>
                  <button
                    type="button"
                    className={css.playlist__play_now}
                    onClick={() => handleSetPlaying(i)}
                  >
                    <Image src={play} alt="play now" />
                  </button>
                  <button
                    type="button"
                    className={css.playlist__remove}
                    onClick={() => removeFromPlaylist(i)}
                  >
                    x
                  </button>
                </div>
              </div>
            )
          })}
        </div>
      )}
      <button
        type="button"
        onClick={() => setShowPlaylist((prev) => !prev)}
        className={css.toggle_playlist}
      >
        <Image src={menu} alt="playlist" height={30} width={30} />
      </button>
    </>
  )
}
