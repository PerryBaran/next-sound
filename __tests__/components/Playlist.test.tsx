import { fireEvent, render, screen } from "@testing-library/react"
import { createContext, useContext } from "react"
import Playlist from "../../src/components/musicPlayer/playlist/Playlist"
import * as PlaylistContext from "../../src/context/PlaylistContext"

describe("Playlist", () => {
  const mockPlaylist = [
    {
      image: "/image-1",
      artistName: "artist-1",
      albumName: "album-1",
      songName: "song-1",
      audio: "audio-1",
      key: "key-1"
    },
    {
      artistName: "artist-2",
      albumName: "album-2",
      songName: "song-2",
      audio: "audio-2",
      key: "key-2"
    }
  ]
  let mockSkipSong = jest.fn()
  let mockRemoveFromPlaylist = jest.fn()
  let mockHandlePlaying = jest.fn()

  afterEach(() => {
    jest.clearAllMocks()
  })

  beforeEach(() => {
    mockSkipSong = jest.fn()
    mockRemoveFromPlaylist = jest.fn()
    mockHandlePlaying = jest.fn()
  })

  describe("playlist not empty", () => {
    beforeEach(() => {
      const mockContext = createContext({
        playlist: mockPlaylist,
        playlistIndex: 0,
        skipSong: mockSkipSong,
        handleAddToPlaylist: jest.fn(),
        removeFromPlaylist: mockRemoveFromPlaylist,
        playing: false,
        handlePlaying: mockHandlePlaying,
        handleShuffle: jest.fn()
      })

      jest
        .spyOn(PlaylistContext, "usePlaylistContext")
        .mockImplementation(() => useContext(mockContext))
    })

    test("renders correctly", () => {
      render(<Playlist />)

      expect(screen.getByRole("button")).toBeInTheDocument()
      expect(screen.getByAltText("playlist")).toBeInTheDocument()
    })

    test("button click renders playlist", () => {
      render(<Playlist />)
      fireEvent.click(screen.getByAltText("playlist"))

      expect(screen.getAllByText(/x/i)).toHaveLength(mockPlaylist.length)
      expect(screen.getAllByAltText(/play now/i)).toHaveLength(
        mockPlaylist.length
      )

      mockPlaylist.forEach((song) => {
        expect(
          screen.getByAltText(`${song.albumName} cover art`)
        ).toBeInTheDocument()
        expect(screen.getByText(song.artistName)).toBeInTheDocument()
        expect(screen.getByText(song.artistName)).toBeInTheDocument()
      })
    })

    test("play now click", () => {
      render(<Playlist />)
      fireEvent.click(screen.getByAltText("playlist"))

      mockPlaylist.forEach((song, i) => {
        fireEvent.click(screen.getAllByAltText("play now")[i])

        expect(mockHandlePlaying).toBeCalledTimes(i + 1)
        expect(mockHandlePlaying).toHaveBeenLastCalledWith(true)
        expect(mockSkipSong).toBeCalledTimes(i + 1)
        expect(mockSkipSong).toHaveBeenLastCalledWith(i)
      })
    })

    test("x click", () => {
      render(<Playlist />)
      fireEvent.click(screen.getByAltText("playlist"))

      mockPlaylist.forEach((song, i) => {
        fireEvent.click(screen.getAllByText("x")[i])

        expect(mockRemoveFromPlaylist).toBeCalledTimes(i + 1)
        expect(mockRemoveFromPlaylist).toHaveBeenLastCalledWith(i)
      })
    })

    test("snapshot", () => {
      const { asFragment } = render(<Playlist />)

      expect(asFragment()).toMatchSnapshot()
    })
  })

  test("empty playlist renders correctly", () => {
    const mockContext = createContext({
      playlist: [],
      playlistIndex: 0,
      skipSong: mockSkipSong,
      handleAddToPlaylist: jest.fn(),
      removeFromPlaylist: mockRemoveFromPlaylist,
      playing: false,
      handlePlaying: mockHandlePlaying,
      handleShuffle: jest.fn()
    })

    jest
      .spyOn(PlaylistContext, "usePlaylistContext")
      .mockImplementation(() => useContext(mockContext))

    render(<Playlist />)
    fireEvent.click(screen.getByAltText("playlist"))

    expect(screen.getByText("Playlist Empty")).toBeInTheDocument()
  })
})
