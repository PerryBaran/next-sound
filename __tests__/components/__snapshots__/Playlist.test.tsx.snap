// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Playlist snapshot 1`] = `
<DocumentFragment>
  <button
    class="toggle_playlist"
    type="button"
  >
    <img
      alt="playlist"
      data-nimg="1"
      decoding="async"
      height="30"
      loading="lazy"
      src="/_next/image?url=%2Fimg.jpg&w=64&q=75"
      srcset="/_next/image?url=%2Fimg.jpg&w=32&q=75 1x, /_next/image?url=%2Fimg.jpg&w=64&q=75 2x"
      style="color: transparent;"
      width="30"
    />
  </button>
</DocumentFragment>
`;
